network/secondary_endpoint_client_linux.go

```diff
@@ -0,0 +1,128 @@
+	endpoint *endpoint,
+) *SecondaryEndpointClient {
+	client := &SecondaryEndpointClient{
+		netlink:        nl,
+		netioshim:      &netio.NetIO{},
+		plClient:       plc,
+		netUtilsClient: networkutils.NewNetworkUtils(nl, plc),
+		ep:             endpoint,
+	}
+
+	return client
+}
+
+func (client *SecondaryEndpointClient) AddEndpoints(epInfo *EndpointInfo) error {
+	iface, err := client.netioshim.GetNetworkInterfaceByMac(epInfo.MacAddress)
+	if err != nil {
+		return newErrorSecondaryEndpointClient(err)
+	}
+
+	epInfo.IfName = iface.Name
+	if _, exists := client.ep.SecondaryInterfaces[iface.Name]; exists {
+		return newErrorSecondaryEndpointClient(errors.New(iface.Name + " already exists"))
+	}
+	client.ep.SecondaryInterfaces[iface.Name] = &InterfaceInfo{
+		Name:              iface.Name,
+		MacAddress:        epInfo.MacAddress,
+		IPAddress:         epInfo.IPAddresses,
+		NICType:           epInfo.NICType,
+		SkipDefaultRoutes: epInfo.SkipDefaultRoutes,
+	}
+
+	return nil
+}
+
+func (client *SecondaryEndpointClient) AddEndpointRules(_ *EndpointInfo) error {
+	return nil
+}
+
+func (client *SecondaryEndpointClient) DeleteEndpointRules(_ *endpoint) {
+}
+
+func (client *SecondaryEndpointClient) MoveEndpointsToContainerNS(epInfo *EndpointInfo, nsID uintptr) error {
+	// Move the container interface to container's network namespace.
+	log.Printf("[net] Setting link %v netns %v.", epInfo.IfName, epInfo.NetNsPath)
+	if err := client.netlink.SetLinkNetNs(epInfo.IfName, nsID); err != nil {
+		return newErrorSecondaryEndpointClient(err)
+	}
+
+	return nil
+}
+
+func (client *SecondaryEndpointClient) SetupContainerInterfaces(epInfo *EndpointInfo) error {
+	log.Printf("[net] Setting link %v state up.", epInfo.IfName)
+	if err := client.netlink.SetLinkState(epInfo.IfName, true); err != nil {
+		return newErrorSecondaryEndpointClient(err)
+	}
+
+	return nil
+}
+
+func (client *SecondaryEndpointClient) ConfigureContainerInterfacesAndRoutes(epInfo *EndpointInfo) error {
+	if err := client.netUtilsClient.AssignIPToInterface(epInfo.IfName, epInfo.IPAddresses); err != nil {
+		return newErrorSecondaryEndpointClient(err)
+	}
+
+	ifInfo, exists := client.ep.SecondaryInterfaces[epInfo.IfName]
+	if !exists {
+		return newErrorSecondaryEndpointClient(errors.New(epInfo.IfName + " does not exist"))
+	}
+
+	if len(epInfo.Routes) < 1 {
+		return newErrorSecondaryEndpointClient(errors.New("routes expected for " + epInfo.IfName))
+	}
+
+	if err := addRoutes(client.netlink, client.netioshim, epInfo.IfName, epInfo.Routes); err != nil {
+		return newErrorSecondaryEndpointClient(err)
+	}
+
+	ifInfo.Routes = append(ifInfo.Routes, epInfo.Routes...)
+
+	return nil
+}
+
+func (client *SecondaryEndpointClient) DeleteEndpoints(ep *endpoint) error {
+	vmns, err := netns.New().Get()
+	if err != nil {
+		return newErrorSecondaryEndpointClient(err)
+	}
+
+	for iface := range ep.SecondaryInterfaces {
+		if err := client.netlink.SetLinkNetNs(iface, uintptr(vmns)); err != nil {
```

Comment:
This should be done from container network namespace

Reply:
updated